packages/react-aria-components/src/DropZone.tsx

```diff
@@ -0,0 +1,178 @@
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaLabelingProps, PressEvents} from '@react-types/shared';
+import {ButtonContext} from './Button';
+import {ContextValue, Provider, RenderProps, SlotProps, useContextProps, useRenderProps} from './utils';
+import {DropOptions, mergeProps, useButton, useClipboard, useDrop, useFocusRing, useHover, useId, usePress, VisuallyHidden} from 'react-aria';
+import {filterDOMProps, useHasChild} from '@react-aria/utils';
+import {InputContext} from './Input';
+import {isVirtualDragging} from '@react-aria/dnd';
+import {LinkContext} from './Link';
+import React, {createContext, ForwardedRef, forwardRef, useMemo, useRef} from 'react';
+import {TextContext} from './Text';
+import {useLabels} from '@react-aria/utils';
+
+export interface DropZoneRenderProps {
+  /**
+   * Whether the dropzone is currently hovered with a mouse.
+   * @selector [data-hovered]
+   */
+  isHovered: boolean,
+  /**
+   * Whether the dropzone is focused, either via a mouse or keyboard.
+   * @selector :focus
+   */
+  isFocused: boolean,
+  /**
+   * Whether the dropzone is keyboard focused.
+   * @selector [data-focus-visible]
+   */
+  isFocusVisible: boolean,
+  /**
+   * Whether the dropzone is the drop target.
+   * @selector [data-drop-target]
+   */
+  isDropTarget: boolean,
+  /**
+   * Whether the button is currently in a pressed state.
+   * @selector [data-pressed]
+   */
+  isPressed: boolean
+}
+// note: possibly add isDisabled prop in the future
+export interface DropZoneProps extends Omit<DropOptions, 'getDropOperationForPoint' | 'onInsert' | 'onRootDrop' | 'onItemDrop' | 'onReorder'>, RenderProps<DropZoneRenderProps>, SlotProps, AriaLabelingProps, PressEvents {}
+
+export const DropZoneContext = createContext<ContextValue<DropZoneProps, HTMLDivElement>>(null);
+
+function DropZone(props: DropZoneProps, ref: ForwardedRef<HTMLDivElement>) {
+  [props, ref] = useContextProps(props, ref, DropZoneContext);
+  let {onPress, onPressStart, onPressEnd, onPressChange, onPressUp} = props;
+  let {dropProps, isDropTarget} = useDrop({...props, ref});
+  let {hoverProps, isHovered} = useHover({});
+  let {focusProps, isFocused, isFocusVisible} = useFocusRing({within: true});
+
+  let buttonRef = useRef<HTMLButtonElement>(null);
+  let inputRef = useRef<HTMLInputElement>(null);
+  let hasButton = useHasChild('button[slot=file]', ref);
+  let hasLink = useHasChild('span[slot=file]', ref);
+
+  let hasInput = useHasChild('input[type=file]', ref);
```

Comment:
Related to my above comment, I would do these as something like this:

```js
let [buttonRef, hasButton] = useSlot();
```

Then you can pass `buttonRef` to the ButtonContext via the `ref` property. You shouldn't need `slot` then, and it would allow users to provide their own `slot` prop to e.g. `<Input>`.